CephFS: prevent hanging NodeGetVolumeStats on `stat()` syscall when an MDS is slow

nixpanic opened the pull request:
When an MDS is slow, the `NodeGetVolumeStats` CSI procedure may hang for a while. In case the MDS does not respond at all (or the reply is lost), the `stat()` syscall may hang indefinitely.

Because Kubelet retries the `NodeGetVolumeStats` CSI procedure regularly, the number of hanging `stat()` syscalls can pike up. Each hanging call will consume a thread (go routine), and this may eventually starve the processing capabilities of the CephFS CSI-nodeplugin container.

The commits in this PR come from #4125, which is planned to have a little more features (like a separate `data/` directory for the contents of the volume).

## Manual Testing

1. create a PVC and start a Pod that uses it
2. let it run for a few minutes, confirm that `NodeGetVolumeStats` calls are in the CephFS node-plugin logs
3. scale down the MDS(s), to simulate a failure
4. check the CephFS node-plugin logs and see if there are unhealthy messages for `NodeGetVolumeStats` calls

Results (see `GRPC response`):
```
I1024 12:31:13.033160       1 utils.go:195] ID: 11 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:31:13.033195       1 utils.go:206] ID: 11 GRPC request: {}
I1024 12:31:13.033274       1 utils.go:212] ID: 11 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:31:13.034119       1 utils.go:195] ID: 12 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:31:13.034261       1 utils.go:206] ID: 12 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:31:13.035323       1 utils.go:212] ID: 12 GRPC response: {"usage":[{"available":1073741824,"total":1073741824,"unit":1}],"volume_condition":{"message":"volume is in a healthy condition"}}

I1024 12:32:33.878002       1 utils.go:195] ID: 13 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:32:33.878281       1 utils.go:206] ID: 13 GRPC request: {}
I1024 12:32:33.878379       1 utils.go:212] ID: 13 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:32:33.879443       1 utils.go:195] ID: 14 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:32:33.879502       1 utils.go:206] ID: 14 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:32:33.879549       1 utils.go:212] ID: 14 GRPC response: {"volume_condition":{"abnormal":true,"message":"health-check has not responded for 119.215949 seconds"}}
I1024 12:34:17.335423       1 utils.go:195] ID: 15 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:34:17.335461       1 utils.go:206] ID: 15 GRPC request: {}
I1024 12:34:17.335542       1 utils.go:212] ID: 15 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:34:17.336271       1 utils.go:195] ID: 16 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:34:17.336350       1 utils.go:206] ID: 16 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:34:17.336442       1 utils.go:212] ID: 16 GRPC response: {"volume_condition":{"abnormal":true,"message":"health-check has not responded for 222.672824 seconds"}}
I1024 12:35:55.149537       1 utils.go:195] ID: 17 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:35:55.149583       1 utils.go:206] ID: 17 GRPC request: {}
I1024 12:35:55.149667       1 utils.go:212] ID: 17 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:35:55.150294       1 utils.go:195] ID: 18 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:35:55.150331       1 utils.go:206] ID: 18 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:35:55.150431       1 utils.go:212] ID: 18 GRPC response: {"volume_condition":{"abnormal":true,"message":"health-check has not responded for 320.486780 seconds"}}
```

Scale up the MDS(s) again, see that the volume becomes healthy again:
```
I1024 12:39:03.241203       1 utils.go:195] ID: 21 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:39:03.241236       1 utils.go:206] ID: 21 GRPC request: {}
I1024 12:39:03.241308       1 utils.go:212] ID: 21 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:39:03.242094       1 utils.go:195] ID: 22 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:39:03.242132       1 utils.go:206] ID: 22 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:39:03.242220       1 utils.go:212] ID: 22 GRPC response: {"volume_condition":{"abnormal":true,"message":"health-check has not responded for 448.545782 seconds"}}
I1024 12:40:40.223909       1 utils.go:195] ID: 23 GRPC call: /csi.v1.Node/NodeGetCapabilities
I1024 12:40:40.223962       1 utils.go:206] ID: 23 GRPC request: {}
I1024 12:40:40.224044       1 utils.go:212] ID: 23 GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}}]}
I1024 12:40:40.224870       1 utils.go:195] ID: 24 GRPC call: /csi.v1.Node/NodeGetVolumeStats
I1024 12:40:40.224989       1 utils.go:206] ID: 24 GRPC request: {"volume_id":"0001-0011-openshift-storage-0000000000000001-af951451-f063-4a67-bf7c-8e788bdc587c","volume_path":"/var/lib/kubelet/pods/b2a6ae50-16a9-4326-83d5-58aafcb70f90/volumes/kubernetes.io~csi/pvc-66c3bfe6-1046-438b-8070-60b070b71149/mount"}
I1024 12:40:40.225807       1 utils.go:212] ID: 24 GRPC response: {"usage":[{"available":1073741824,"total":1073741824,"unit":1}],"volume_condition":{"message":"volume is in a healthy condition"}}
```

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic committed:
util: add health-checker for periodic filesystem checks

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
cephfs: enable VolumeCondition with new health-checker

The HealthChecker is configured to use the Staging path pf the volume,
with a `.csi/` subdirectory. In the future this directory could be a
directory that is not under the Published directory.

Fixes: #4219
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
doc: Add initial design notes for the Health Checker

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
util: if FilesystemNodeGetVolumeStats succeeds the volume is healthy

When FilesystemNodeGetVolumeStats() succeeds, the volume must be
healthy. This can be included in the VolumeCondition CSI message by
default.

Checks that detect an abnormal VolumeCondition should prevent calling
FilesystemNodeGetVolumeStats() as it is possible that the function will
hang.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -711,8 +744,38 @@ func (ns *NodeServer) NodeGetVolumeStats(
+
+func getHealthCheckPath(stagingTargetPath string) (string, error) {
+	dir := path.Join(stagingTargetPath, ".csi")
+	err := os.Mkdir(dir, 0o700)
```
`os.Mkdir()` or any other check that the directory on the CephFS mountpoint exists is problematic. The NodeGetVolumeStats call hangs when there is no available MDS (to cause intentional failures).

The directory for the timestamp should be created/checked once, when the health-checker is started (just after mounting, likely to succeed).

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -270,6 +272,9 @@ func (ns *NodeServer) NodeStageVolume(
 		}
 	}
 
+	err = ns.healthChecker.StartChecker(req.GetVolumeId(), stagingTargetPath)
+	log.WarningLog(ctx, "failed to start healtchecker: %v", err)
+
```
What if the volume is already mounted (case where csi plugin just restarted)

nixpanic replied:
In that case, there it no health-checker.

Possibly `NodeGetVolumeStats` can be extended to start the checker to catch this.

Madhu-1 replied:
I think we need this , can we have a tracker for this one?

nixpanic replied:
yes, makes sense. #4219 has been created for this.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -599,6 +606,9 @@ func (ns *NodeServer) NodeUnstageVolume(
 
 	volID := req.GetVolumeId()
+
+	ns.healthChecker.StopChecker(volID)
```
This should happen right before umount but after all other operations right?

nixpanic replied:
It should be done early, yes. Unmounting may not be possible if the checker writes/reads from the mountpoint at the same time.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -694,6 +711,18 @@ func (ns *NodeServer) NodeGetVolumeStats(
 	}
 
+	// health check first, return without stats if unhealthy
+	healthy, msg := ns.healthChecker.IsHealthy(req.GetVolumeId())
```
what happens if the NodeGetVolumeStats is called right after the NodeStageVolume?

nixpanic replied:
Before running the 1st check, the volume is regarded healthy. It is the assumption that when mounting works, the volume is still healthy shortly after.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -711,7 +740,12 @@ func (ns *NodeServer) NodeGetVolumeStats(
 	if stat.Mode().IsDir() {
-		return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath, false)
+		res, err := csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath, false)
+		if err != nil {
+			return nil, err
+		}
+
+		return res, nil
```
we dont need this check and can have `return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath, false)` as it was before

nixpanic replied:
ah, yes. It was modified in an earlier version to set `res.VolumeCondition = ...` after getting the stats. ... that wasn't my smartest move, as getting the stats got hung, so the `VolumeCondition` was never set and returned :man_facepalming:

nixpanic replied:
Done.

Madhu-1 commented on internal/health-checker/filechecker.go:
```diff
@@ -0,0 +1,186 @@
+	"time"
+)
+
+// command is what is sent through the channel to terminate the go routine.
+type command string
+
+const (
+	// stopCommand is sent through the channel to stop checking.
+	stopCommand = command("STOP")
+)
```
can make use of empty struct to signal between go routines?

nixpanic replied:
That is possible, but what is the advantage?

I've chosen this way to make it extendable if needed, it does not matter at the moment what is sent over the channel.

Madhu-1 replied:
In most cases empty structs are preferred to signal in the go routines, that's why the suggestion

nixpanic replied:
Is that a recommendation that is documented anywhere? I have seen several examples with integers and booleans, or structs that contain one or more members...

Madhu-1 commented on internal/health-checker/filechecker.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+func (fc *fileChecker) shouldStop() bool {
+	start := time.Now()
+
+	for {
+		// check if we slept long enough to run a next check
+		slept := time.Since(start)
+		if slept >= fc.interval {
+			break
+		}
+
+		select {
+		case <-fc.commands:
+			// a command was reveived, need to stop checking
+			return true
+		default:
+			// continue with checking
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return false
```
use time.Ticker for this one?

nixpanic replied:
yes, I'll do, that makes it a little cleaner.

nixpanic replied:
Done.

Madhu-1 commented on internal/health-checker/filechecker.go:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	//nolint:gosec // allow reading of the timestamp for debugging
+	return os.WriteFile(fc.filename, data, 0o644)
```
will this also get blocked like os.Stat as well or we will receive an error?

nixpanic replied:
Yes. The behavior for all (I guess) syscalls will be the same. Either blocking, or returning an unexpected error (like permission denied after fencing).

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -694,6 +711,18 @@ func (ns *NodeServer) NodeGetVolumeStats(
+	// health check first, return without stats if unhealthy
+	healthy, msg := ns.healthChecker.IsHealthy(req.GetVolumeId())
+	if !healthy {
+		return &csi.NodeGetVolumeStatsResponse{
```
What happens for the already running workload if cephcsi is upgraded for this one version?

nixpanic replied:
Nothing. The `VolumeCondition` is not used anywhere directly. [Kubernetes can report it](https://kubernetes.io/docs/concepts/storage/volume-health-monitoring/), but does not do so by default (yet).

Madhu-1 replied:
what i mean was if the cephcsi is upgraded we might return the volume as unhealthy for the already mounted pvc because health checker will only gets started during NodeStageVolume? same question for cephcsi restart as well. `IsHealthy` returns error if there is no health check found

nixpanic replied:
Right, tracked in #4219.

nixpanic replied:
this is actually addressed in the latest revision of this PR too

Madhu-1 commented on docs/design/proposals/volume-condition.md:
```diff
@@ -0,0 +1,25 @@
+The health-checker writes to the file `csi-volume-condition.ts` in the root of
+the volume. This file containse a JSON formatted timestamp.
+
+A new `data` directory is introduced for newly created volumes. During the
+`NodeStageVolume` call the root of the volume is mounted, and the `data`
```
Will there be any problem for the static PVC or the existing PVC which already have data folder created in it? i don't think so but checking

nixpanic replied:
Existing PVC's would not have the option in the VolumeContext, so there will be a fall-back mechanism.

The current implementation uses a hidden `.csi/` directory in the root of the volume, which is basically what is used for volumes that do not have the extra (coming) option in the VolumeContext.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -270,6 +272,9 @@ func (ns *NodeServer) NodeStageVolume(
 		}
 	}
 
+	err = ns.healthChecker.StartChecker(req.GetVolumeId(), stagingTargetPath)
```
I might be missing the code where We create the data directory and mount that to the target path from stagingPath.

nixpanic replied:
That is not part of this PR. This only addresses the async health-checking. #4125 was started with the extra `data` subdir in mind, and it will be remodelled to extend this PR.

Madhu-1 replied:
The same check needs to be done at line 233 where we are returning success if the directory is already mounted

nixpanic replied:
Done.

Madhu-1 commented on internal/health-checker/filechecker.go:
```diff
@@ -0,0 +1,186 @@
+
+	delay := time.Since(fc.lastUpdate)
+	if delay > (fc.interval + fc.timeout) {
+		fc.healthy = false
+		fc.err = fmt.Errorf("health-check has not responded for %f seconds", delay.Seconds())
```
Will there be any problem with multiple goroutines trying to update the same variables?

nixpanic replied:
Not really. There is a very small race in case one go-routine does not enter this if-statement, and an other does + updates the `fc.err` before the 1st go-routines exits the function. I don't think it is a realistic failure scenario to account for.

A same way can happen with this function and concurrent `runChecker()`. But it also looks quite an edge case.

If you insist, I can add some locking around updating/reading `fc.healthy` and `fc.err`.

nixpanic replied:
Locking has been added.

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+		return fmt.Errorf("failed to created workdir %q for health-checker: %w", workdir, err)
+	}
+
+	cc := newFileChecker(workdir)
```
This always starts a FileChecker what about BlockChecker for future?

nixpanic replied:
When a `BlockChecker` is introduced, the `Manager` interface will need some modifications in any case.

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+		if !ok {
+			return fmt.Errorf("failed to cast cc to ConditionChecker for volume-id %q", volumeID)
+		}
+	} else {
+		// 'cc' was stored, start it only once
+		cc.start()
+	}
```
This can happen only for Static PVC? because the NodeStageVolume will be called only once per volumeID

nixpanic replied:
`NodeStageVolume` might have failed/timedout after the checker was started. Calling `NodeStageVolume` again should be a no-op in that case, this just ensures some idempotency.

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+
+	cc := newFileChecker(workdir)
+
+	// load the 'old' ConditionChecker if it exists, otherwuse store 'cc'
```
`otherwuse` to `otherwise`

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+	cc.stop()
+}
+
+func (hcm *healthCheckManager) IsHealthy(volumeID string) (bool, error) {
```
How manager behaves when the same volumeID is used for both cephfs/NFS and rbd?

nixpanic replied:
That should not matter, this Manager runs in a node-plugin, and those are dedicated for cephfs/rbd/nfs.

I am also not sure how you could get the same VolumeID for different storage backends?

riya-singhal31 commented on docs/design/proposals/volume-condition.md:
```diff
@@ -0,0 +1,25 @@
+the (un)healthy state and stop the checking process when the volume is not
+needed anymore.
+
+The Manager is responsible for creating a suitble checker for the requested
```
minor nits,  suitable

riya-singhal31 commented on docs/design/proposals/volume-condition.md:
```diff
@@ -0,0 +1,25 @@
+needed anymore.
+
+The Manager is responsible for creating a suitble checker for the requested
+path. If the path a s block-device, the BlockChecker should be created. For a
```
```suggestion
path. If the path is block-device, the BlockChecker should be created. For a
```

riya-singhal31 commented on docs/design/proposals/volume-condition.md:
```diff
@@ -0,0 +1,25 @@
+## CephFS
+
+The health-checker writes to the file `csi-volume-condition.ts` in the root of
+the volume. This file containse a JSON formatted timestamp.
```
```suggestion
the volume. This file contains a JSON formatted timestamp.
```

nixpanic commented:
@Madhu-1 and @riya-singhal31 , I have manually tested this again (by scaling MDS's down/up) and it still works as before. Could you review again and mark the addressed comments as resolved? Many thanks!

riya-singhal31 commented on internal/cephfs/nodeserver.go:
```diff
@@ -270,6 +272,9 @@ func (ns *NodeServer) NodeStageVolume(
 	}
 
+	err = ns.healthChecker.StartChecker(req.GetVolumeId(), stagingTargetPath)
+	log.WarningLog(ctx, "failed to start healtchecker: %v", err)
```
```suggestion
	log.WarningLog(ctx, "failed to start healthchecker: %v", err)
```

nixpanic replied:
Done.

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (hcm *healthCheckManager) StartChecker(volumeID, path string) error {
+	workdir := filepath.Join(path, ".csi")
```
what if the PVC already has this path or the application doesn't expect any folders inside the mounted directory?

nixpanic replied:
Other filesystems also maintain their own metadata in the root, like `lost+found` on ext4. Applications that (do not) expect certain filesystem internal metadata should be quite rare, as they are not very portable.

Would you prefer to have an option per volume to use a different directory, or disable it?

Madhu-1 replied:
Yes, I agree on that one, this is something we are introducing, but not all the application owners/developers will be aware of it. IMHO we should not write or touch anything in the path that is provided to the application pods, I would prefer the below

* Mount the subvolume to the stagingPath and create a new directory as in the design doc and mount that directory to the targetPath
* Do health check only inside the stagingPath which is not available to the application pod

nixpanic replied:
For legacy volumes, we could also do a minimal `stat()` on the root of the volume, instead of writing/reading a timestamp. It might not be able to detect all failures scenarios, but it would be a non-intrusive check.

Newly created volumes can have their own `data` subdirectory that is mapped for use inside of the container, leaving the root of the volume for contents that is not visible to applications. That can be done in PR #4125 once this one is accepted.

Madhu-1 replied:
>For legacy volumes, we could also do a minimal stat() on the root of the volume, instead of writing/reading a timestamp. It might not be able to detect all failures scenarios, but it would be a non-intrusive check.

It's a temporary problem, once the node gets rebooted or the application gets restagted we will move to the new format. We can live with it.

>Newly created volumes can have their own data subdirectory that is mapped for use inside of the container, leaving the root of the volume for contents that is not visible to applications. That can be done in PR https://github.com/ceph/ceph-csi/pull/4125 once this one is accepted.

Sounds good 👍🏻

nixpanic replied:
Added `statChecker` and will use that in the nodeserver for now. `fileChecker` is still included so that the refactoring with a base `checker` makes sense.

Madhu-1 reviewed (commented):
Left some questions, Mostly LGTM

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -270,6 +275,9 @@ func (ns *NodeServer) NodeStageVolume(
 	}
 
+	err = ns.healthChecker.StartChecker(req.GetVolumeId(), stagingTargetPath)
+	log.WarningLog(ctx, "failed to start healthchecker: %v", err)
```
this should be done inside `if err !=nil` check

nixpanic replied:
done!

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -228,6 +230,9 @@ func (ns *NodeServer) NodeStageVolume(
 		}
 
+		err = ns.healthChecker.StartChecker(req.GetVolumeId(), stagingTargetPath)
+		log.WarningLog(ctx, "failed to start healthchecker: %v", err)
```
this should be done inside `if err !=nil` check

nixpanic replied:
Done!

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+	// load the 'old' ConditionChecker if it exists
+	old, ok := hcm.checkers.Load(volumeID)
+	if !ok {
+		return true, fmt.Errorf("no ConditionChecker for volume-id: %s", volumeID)
```
This error will not checked  as we are returning volume as Healthy and also the error message

nixpanic replied:
Added a check for that now, will start a checker in that case too.

Madhu-1 commented on internal/health-checker/manager.go:
```diff
@@ -0,0 +1,121 @@
+	// 'old' was loaded, cast it to ConditionChecker
+	cc, ok := old.(ConditionChecker)
+	if !ok {
+		return true, fmt.Errorf("failed to cast cc to ConditionChecker for volume-id %q", volumeID)
```
This error will not checked as we are returning volume as Healthy and also the error message

nixpanic replied:
If there is an error in the internals of the health-checker, it is assumed that the volume is healthy. Only when there really is a problem with the volume, the status is reported as unhealthy.

Madhu-1 replied:
Can we log and also document this one? we might need to have better logging to check whether the health checker is running and what was the latest run details (by looking at the logs), maybe we can do it as an enhancement later.

nixpanic replied:
I've added more docs about it now.

nixpanic commented:
Lots of manual testing done. including the following combinations:

pod uses PVC and

1. is healthy
2. csi-cephfsplugin pods restarted, volume stays healthy
3. MDS' are stopped, volume becomes unhealthy
4. MDS' restarted, volume becomes healthy again
5. csi-cephfsplugin pods restarted when MDS' are unavailable, volume stays unhealthy
6. csi-cephfsplugin restarted while unhealthy, MDS' restarted, volume becomes healthy again

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -694,6 +728,34 @@ func (ns *NodeServer) NodeGetVolumeStats(
+	// volume is already staged and published.
+	if healthy && msg != nil {
+		// Start a StatChecker for the mounted targetPath, yhis prevents
+		// writing a file in the user-visible location. Ideally a (shared)
```
`yhis` to `this`

riya-singhal31 reviewed (dismissed):
Thanks Niels, LGTM

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -694,6 +728,34 @@ func (ns *NodeServer) NodeGetVolumeStats(
+		// writing a file in the user-visible location. Ideally a (shared)
+		// FileChecker is started with the stagingTargetPath, but we can't
+		// get the stagingPath from the request.
+		err = ns.healthChecker.StartChecker(req.GetVolumeId(), targetPath, hc.StatCheckerType)
```
can we use something like below to get the stagingPath from targetPath 
https://github.com/ceph/ceph-csi/blob/c09700b5688847cfa2efad0ae1dab5cc14e521ed/internal/rbd/nodeserver.go#L1079-L1094

nixpanic replied:
Something like this could be an enhancement. I am not sure if it is worth the effort and additional complexity though.

When a FileChecker is used, it would make more sense as that could detect the existence of the file holding the timestamp.

nixpanic replied:
Adding a TODO comment with reference to `rbd.getStagingPath()`

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -694,6 +728,34 @@ func (ns *NodeServer) NodeGetVolumeStats(
+		// writing a file in the user-visible location. Ideally a (shared)
+		// FileChecker is started with the stagingTargetPath, but we can't
+		// get the stagingPath from the request.
+		err = ns.healthChecker.StartChecker(req.GetVolumeId(), targetPath, hc.StatCheckerType)
```
i also see a very small race for the below case

* During NodeUnpublish cephcsi stopped the health checker
* Before umount operation we got a `NodeGetVolumeStats` RPC call which inturn started a health checker
* Now we have a stale health check which will be cleanedup only during NodeUnstage (i think)

Is it possible to have such a case?

nixpanic replied:
I do not expect the CO to request `NodeUnpublish` and then still use the same `publishTargetPath` in a later `NodeGetVolumeStats`. If a CO does that, then yes, there would be a stale health checker (if it was started in a `NodeGetVolumeStats` call, and not in `NodeStageVolume`).

Madhu-1 reviewed (commented):
LGTM, left a question and small nit

nixpanic commented:
@Madhu-1 I have addressed your latest concerns as well. Please check again, thanks!

Madhu-1 reviewed (approved):
LGTM, Thanks @nixpanic

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue
